### ISSUE TITLE
User certificate password is not a parameter

### ISSUE THREAD

**Issue by alfem** [adrienverge/openfortivpn#840]
My organization uses personal certs for vpn connections. 

These certs are password protected, and openfortivpn ask for the cert key password twice before establishing the connection and once more to close it (weird!).

AFAIK there is no command line or configuration file parameter for this password, so the only option is saving your key file without password. 

As this personal certificate is provided by our government is not easy to obtain new one, and most people do not want to store it without a password. 

It would be nice to have this parameter added, or at least make openfortivpn ask for the password only once.

**Comment by DimitriPapadopoulos** [comment]
I don't have a certificate myself, so I'm afraid I won't be much help. Perhaps _openfortivpn_ asks once for an (empty) password and once for the certificate password itself. Do you really have to enter the certificate password twice, or entering an empty password once works too?

This of course doesn't explain why a password is required to close the connection, but let's leave this issue aside for now.

A (sanitized) log might help identify the problem:
```
openfortivpn -v -v
```

**Comment by DimitriPapadopoulos** [comment]
And by the way, which version of _openfortivpn_ is this?

**Comment by DimitriPapadopoulos** [comment]
I had already noticed that within _run_tunnel_(), _ssl_connect_() is called twice when establishing the VPN connection:
https://github.com/adrienverge/openfortivpn/blob/ba44ce1/src/tunnel.c#L1215
https://github.com/adrienverge/openfortivpn/blob/ba44ce1/src/tunnel.c#L1241
and once when closing the VPN connection:
https://github.com/adrienverge/openfortivpn/blob/ba44ce1/src/tunnel.c#L1293

This had been discussed in #239, but I'm still failing to fully grasp the logic behind these _ssl_connect_()  / _ssl_disconnect_() calls. It might be related. I really don't know, just a shot in the dark.

**Comment by alfem** [comment]
>  Do you really have to enter the certificate password twice, or entering an empty password once works too?
 
Yes, I do. I tried with an empty password and connections fails. Reading #239 I guess openfortivpn first cleans the tunnel closing it (even when it is not open yet). That triggers a first password prompt. Then it asks for the password again in order to open the tunnel properly. I do not know (my C-fu is weak) if caching the password is possible.

> A (sanitized) log might help identify the problem:

`sudo openfortivpn vpn.ERASED:443 -u alfERASED -p ERASED --user-cert=/home/MYUSER/MYCERT-2021.pem --user-key=/home/MYUSER/MYKEY-2021.key  -v -v --trusted-cert 17bdb2d9960bd63dd9469700d4cdb35debb14651dc11879f68a72d740994332e
`

Here you are: https://pastebin.com/68imWA8F

**Comment by DimitriPapadopoulos** [comment]
Please bear with me as I cannot test certificates myself.

From the log, it looks like _openfortivpn_ loads config file `/etc/openfortivpn/config`:
```
DEBUG:  Loaded config file "/etc/openfortivpn/config".
DEBUG:  Config host = "vpn.ERASED"
DEBUG:  Config realm = ""
DEBUG:  Config port = "443"
DEBUG:  Config username = "alfERASED"
DEBUG:  Config password = "********"
```
Also, you pass the username and the password as command line arguments:
```
sudo openfortivpn vpn.ERASED:443 -u alfERASED -p ERASED ...
```
I seem to recall command line arguments take precedence over the config file. Anyway, both in the config file and the command line argument, is the `ERASED` password an empty/random password or the certificate password? My understanding is that it doesn't need to be the certificate password, and that the certificate password is to be entered interactively. But then in the log I don't see the prompt that asks for the certificate password interactively. And I'm not certain where the _openfortivpn_ code that prompts the certificate password is. I can find only 3 calls to _read_password_() that are related to regular passwords, not to certificate passwords:
* https://github.com/adrienverge/openfortivpn/blob/ba44ce1/src/main.c#L603 (prompt `VPN account password: `)
* https://github.com/adrienverge/openfortivpn/blob/ba44ce1/src/http.c#L586 (prompt `Please enter one-time password:`)
* https://github.com/adrienverge/openfortivpn/blob/ba44ce1/src/http.c#L731 (prompt `Two-factor authentication token: `)

What does the prompt for the certificate password look like?

**Comment by alfem** [comment]
> From the log, it looks like _openfortivpn_ loads config file `/etc/openfortivpn/config`:

Not at all. Config file is empty, no matter what the log says. I am only using command line parameters.

Perhaps de DEBUG lines show the parameters after resolving precedence.


> I seem to recall command line arguments take precedence over the config file. Anyway, both in the config file and the command line argument, is the `ERASED` password an empty/random password or the certificate password? 

The `ERASED` password is the vpn password, not the certificate one. I need to provide both auth parameters: user/password and cert/cert-key. 

> My understanding is that it doesn't need to be the certificate password, and that the certificate password is to be entered interactively. 

Yes, that is the way it works now. 

> What does the prompt for the certificate password look like?

Probably it is not shown in the log because I just redirected the standard output. It looks like this:

`Enter PEM pass phrase:`

**Comment by DimitriPapadopoulos** [comment]
I cannot find such a prompt in the _openfortivpn_ sources. I suspect this is _OpenSSL_ prompting for the certificate password. And indeed:
https://github.com/openssl/openssl/blob/9327b5/crypto/pem/pem_lib.c#L49

We must find a way to prompt for this password in _openfortivpn_ and pass it to _OpenSSL_.

**Comment by mrbaseman** [comment]
Yes, I can confirm that this is the prompt from openssl. It has always been like this in openfortivpn. 

I have converted my user certificate and removed the pem passphrase, but introduced the 2fa with the fortitoken app instead as a real second factor. When I was looking at this some time ago, I couldn't figure out how to tell openssl to cache it for the current session.

But well, there is [SSL_CTX_set_default_passwd_cb](https://www.openssl.org/docs/man1.1.1/man3/SSL_CTX_set_default_passwd_cb.html) which sets a callback function that the application has to provide and that hands back the password to be used during decryption. This callback function could

1. use a password stored in the config file
2. otherwise ask the user for it
3. cache it for later use and don't ask when called a second time

**Comment by alfem** [comment]
Yes, you are right. 

A new command line/config file parameter is needed, and this new password should decipher the key file using  [SSL_CTX_set_default_passwd_cb]( https://www.openssl.org/docs/man1.1.1/man3/SSL_CTX_set_default_passwd_cb.html), if I am not wrong.

**Comment by mrbaseman** [comment]
Yes. It seems this has been possible in openssl for a long time, so we don't need to distinguish between different versions of openssl.

**Comment by DimitriPapadopoulos** [comment]
@mrbaseman Would it be possible to create an account that works with a user certificate on the test FortiGate?

**Comment by DimitriPapadopoulos** [comment]
I have prepared a patch. Because I cannot test, I don't know if it works - at least it compiles.

Please note that the OpenSSL documentation is not precise about the length of the buffer: with or without the trailing ``\0'`? Will probably have to investigate by reading the OpenSSL source code.

**Comment by alfem** [comment]
> I have prepared a patch. Because I cannot test, I don't know if it works - at least it compiles.

I could test it, if your provide the name of the branch/fork.

**Comment by mrbaseman** [comment]
I could test it, but unfortunately I got
```
Enter PEM pass phrase: 
ERROR:  SSL_CTX_use_PrivateKey_file: error:140B0009:SSL routines:SSL_CTX_use_PrivateKey_file:PEM lib
INFO:   Closed connection to gateway.
ERROR:  SSL_CTX_use_PrivateKey_file: error:140B0009:SSL routines:SSL_CTX_use_PrivateKey_file:PEM lib
INFO:   Could not log out.
```

**Comment by DimitriPapadopoulos** [comment]
Mmmh... i cannot make anything out of this error message. Perhaps it's an issue with the "buffer length" argument which is not properly defined in the man page (with or without the final `\0`?). I'll probably have to read the OpenSSL source code to make sure. Or perhaps I haven't set the callback properly.

The branch is [DimitriPapadopoulos:pem_passphrase](https://github.com/DimitriPapadopoulos/openfortivpn/tree/pem_passphrase).

**Comment by alfem** [comment]
@DimitriPapadopoulos, looks like the callback is returning the lenght of the passphrase, not the passphrase itself. Is that right?

**Comment by DimitriPapadopoulos** [comment]
Yes, that's my understanding of how the callback works.

**Comment by mrbaseman** [comment]
> @mrbaseman Would it be possible to create an account that works with a user certificate on the test FortiGate?

Hmm... well we would have to add a trusted CA and get a vaild user certificate from that CA. And then I think it's a global configuration for all accounts. 
For testing I just had to put a password on the private key for another appliance that works with user certificates anyway.
